workbench-core/authorization/src/dynamicAuthorization/dynamoDBDynamicPermissionsPlugin.ts

```diff
@@ -0,0 +1,491 @@
+import { QueryCommandOutput, GetItemCommandOutput } from '@aws-sdk/client-dynamodb';
+import { marshall } from '@aws-sdk/util-dynamodb';
+import { AwsService } from '@aws/workbench-core-base';
+import _ from 'lodash';
+import { Action } from '../action';
+import { Effect } from '../permission';
+import { DynamicPermissionsPlugin } from './dynamicPermissionsPlugin';
+import {
+  CreateGroupRequest,
+  CreateGroupResponse,
+  DeleteGroupRequest,
+  DeleteGroupResponse,
+  GetUserGroupsRequest,
+  GetUserGroupsResponse,
+  CreateIdentityPermissionsRequest,
+  CreateIdentityPermissionsResponse,
+  DeleteIdentityPermissionsRequest,
+  DeleteIdentityPermissionsResponse,
+  DeleteSubjectPermissionsRequest,
+  DeleteSubjectPermissionsResponse,
+  AssignUserToGroupRequest,
+  AssignUserToGroupResponse,
+  RemoveUserFromGroupRequest,
+  RemoveUserFromGroupResponse,
+  GetIdentityPermissionsBySubjectRequest,
+  GetIdentityPermissionsBySubjectResponse,
+  IdentityPermission,
+  IdentityType,
+  GetUsersFromGroupRequest,
+  GetUsersFromGroupResponse,
+  GetIdentityPermissionsByIdentityRequest,
+  GetIdentityPermissionsByIdentityResponse
+} from './dynamicPermissionsPluginInputs';
+import { BadConfigurationError } from './errors/badConfigurationError';
+import { GroupAlreadyExistsError } from './errors/groupAlreadyExistsError';
+import { GroupNotFoundError } from './errors/groupNotFoundError';
+
+export interface InitRequest {
+  groupsToBeCreated?: { groupIds: string[] };
+  identityPermissionsToBeCreated?: IdentityPermission[];
+}
+
+export interface InitResponse {
+  success: boolean;
+}
+
+export class DynamoDBDynamicPermissionsPlugin implements DynamicPermissionsPlugin {
+  private _awsService: AwsService;
+  private _initializationPromise: boolean;
+  private _getResourceBySortKeyIndex: string;
+  private readonly _groupPrefix: string = 'groupids';
+
+  private readonly _assignedGroupPrefix: string = 'assigned';
+
+  private readonly _identityPermissionsPrefix: string = 'ip';
+
+  public constructor(params: { awsService: AwsService; getResourceBySortKeyIndex: string }) {
+    this._awsService = params.awsService;
+    this._initializationPromise = false;
+    this._getResourceBySortKeyIndex = params.getResourceBySortKeyIndex;
+  }
+
+  public async init(initRequest: InitRequest): Promise<InitResponse> {
+    try {
+      // Check if table is empty
+      const response = await this._awsService.helpers.ddb.scan().execute();
+      if (response.Count && response.Count > 0) {
+        throw new BadConfigurationError('DynamoDB is not empty');
+      }
```

Comment:
Is it a requirement for the ddb table used here be exclusively for dynamic authZ?

Reply:
Yes, authz data is higher sensitivity and would be preferred to sit separately away from application data